ipa-core/src/helpers/stream/chunks.rs

```diff
@@ -720,22 +812,70 @@ mod tests {
         assert!(poll_immediate(&mut st).next().await.is_none() && st.is_terminated());
     }
 
+    impl<K, const N: usize> Chunk<K, N> {
+        pub fn new(chunk_type: ChunkType, data: K) -> Self {
+            Self { chunk_type, data }
+        }
+    }
+
     #[test]
     fn chunk_into_iter() {
         assert_eq!(
-            Chunk::new(ChunkType::Full, [1, 2])
+            Chunk::<_, 2>::new(ChunkType::Full, [1, 2])
                 .into_iter()
                 .collect::<Vec<_>>(),
             [1, 2],
         );
         assert_eq!(
-            Chunk::new(ChunkType::Partial(1), [3, 4])
+            Chunk::<_, 2>::new(ChunkType::Partial(1), [3, 4])
                 .into_iter()
                 .collect::<Vec<_>>(),
             [3],
         );
     }
 
+    #[test]
+    fn chunk_unpack() {
+        assert_eq!(
+            Chunk::<_, 4>::new(ChunkType::Full, vec![[1, 2], [3, 4]]).unpack::<2>(),
```

Comment:
The `N` const parameter on `Chunk` is the number of records in a fully populated chunk. The `K` type parameter is the storage type (typically something like `[T; N]` or `Vec<T>`).

The `Chunk::unpack` transformer is a special case to support converting between vectorization by 256 for share conversion, and vectorization by 16 for PRF evaluation. These are not the actual types used, but logically, it can be thought of like this: share conversion outputs `Chunk<Vec<Vec<Row>>, 256>`. The inner vectors each have 16 elements, corresponding to PRF vectors. The outer vectors have 256 / 16 = 16 elements. The unpack transform converts `Chunk<Vec<Vec<Row>>, 256>` to `Vec<Chunk<Vec<Row>>, 256>`. PRF evaluation is then invoked on each `Chunk<Vec<Row>, 256>` in the unpacked output `Vec`.

The `chunk_unpack` test is exercising just the unpack transform, verifying that a 4-chunk containing two 2-element sub-chunks can be unpacked into those sub-chunks. That the sub-chunks have length two is passed as the const parameter to the `unpack` method.

For the test, the type `T` (the inner array-ish type) is truly an array. But in the real protocol, it's an `AdditiveShare<Fp25519, NP>`. An impl of `fn unpack` for `Chunk<Vec<[U; M]>, N>` wouldn't work there.

As I [mentioned](https://github.com/private-attribution/ipa/pull/1087#discussion_r1610895698), it would be nice for the output type to be `[Chunk<T, M>; N / M]`, but division of the const parameters isn't currently supported.